File: leetcode-solutions/0020-valid-parentheses/solution.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const stack = [];
    
    for (let i = 0; i < s.length; i++) {
    const ch = s[i];
    
    if (ch === '(' || ch === '{' || ch === '[') {
      stack.push(ch);
    } 
    
    else if (ch === ')' || ch === '}' || ch === ']') {
        const top = stack.pop();
        if (
            (ch === ')' && top !== '(') ||
            (ch === '}' && top !== '{') ||
            (ch === ']' && top !== '[')
            ) 
        {
            return false;
        }
    }
  }

  return stack.length === 0;
};

File: leetcode-solutions/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        lookUp = {'(': ')', '[': ']', '{': '}'}
        stack = []

        for ch in s:
            if ch in lookUp:  # opening
                stack.append(ch)
                
            else:  # closing
                if len(stack) == 0:
                    return False

                open_br = stack.pop()
                if lookUp[open_br] != ch:
                    return False

        return len(stack) == 0

